Clarify service worker timeout (#5096)
GoogleChrome/developer.chrome.com: site/en/blog/longer-esw-lifetimes/index.md

  A picture of an hourglass.
---

{% Aside %}

> This post has been updated since we originally published it in January. TBD -summary of changes.

Author: @sebastianbenz I'll fill this out when the content is settled.

> This post has been updated since we originally published it in January. We've added text indicating that calls to extension APIs reset the service worker's timeout clock, and clarified exactly when service worker timeouts will ocurr.

Reviewer: Hopefully this is the US spelling too:

```suggestion
This post has been updated since we originally published it in January. We've added text indicating that calls to extension APIs reset the service worker's timeout clock, and clarified exactly when service worker timeouts will occur.
```

## What's changed

> As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now typically stay alive as long as they are actively processing events. Be aware that:

Reviewer 2: I think it'd be great to elaborate a bit more on this:

- why are we're making this change?
- what was the previous behavior? (this helps devs gauge how this might impact the behavior of their extension)

Another thing: can we either remove `typically` or make it more precise? left on its own it is a bit confusing (at least I stumbled over it).

Author: >why are we're making this change?

Because premature termination was causing errors in people's code. At the time we figured that if somebody wasn't  having that problem they wouldn't be reading this article. Frankly, I'd like to err on the side of getting this out.

>what was the previous behavior?

It's described in the background.

* The service worker will terminate after 30 seconds if there are no events.

> * The service worker will terminate after 5 minutes if a single request, such as an event or API call, takes longer than 5 minutes to process (with the exception of a native messaging connection).

Reviewer 2: I assume this is only the case for pending async calls, correct? Or will synchronous operations be terminated as well?

//cc @rdcronin

Author: I don't know. Maybe Devlin does.

Author: New wording from Oliver.

Author: Oliver didn't like the new wording and he makes some compelling arguments:
* >It won't terminate after 30 seconds if you make certain API calls for example. So it would need to be "The service worker will terminate after 30 seconds if there are no events or API calls to extension APIs and there isn't a native messaging connection, regardless of whether you're running synchronous or asynchronous code." I realise we touch on that elsewhere in the article but I think we should make sure if people quote a specific bullet point it isn't wrong.
* >I feel like synchronous and asynchronous was sort of implied with how we had it before, and calling it out actually adds confusion because then it makes me stop to think about what those are.

Reviewer: I'm not sure what the latest wording is but I think something like this would be fine:

> * The service worker will terminate after 30 seconds of inactivity (receiving an event or calling an extension API resets this timer).
> * The service worker will terminate after 5 minutes if a single request, such as an event or API call, takes longer than 5 minutes to process
>
> Some APIs like native messaging provide a strong keep-alive which cancel both of these timers.

And maybe change "be aware that" to "specifically" or something? Since it's not a bad thing.

Reviewer 2: +1 to Oliver's suggestions

> You are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally. Make sure your extensions yield when possible and prepare for termination by persisting extension states.

Reviewer 2: This does not quite fit the rest of the flow: "You are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination."

Reading the post I see nothing that would indicate that sw's are no longer event-based (as we don't explicitly mention hacking the lifecycle).

Author: I remember no. Simeon strongly insisted on this line. We can get rid of it. @oliverdunk I'd be curious to know, if you have any opinions on this line.

Reviewer: I think the two important things to communicate are:

- We may change this behaviour to be stricter in the future (though hopefully in a way that does not impact genuine extensions).
- Your extension may terminate unexpectedly even with these changes (for example, if the browser crashes).

How about something like:

> In a future release, we may continue to iterate on this behaviour to ensure extension service workers are terminated when possible, without shutting down long-running work. As part of building a resource conscious extension, extension service workers should always yield when possible. Additionally, extensions should prepare for unexpected termination by persisting state. This makes sure they are robust against unpredictable events like the browser being forcefully closed by the user.

I think that's a bit wordy but better in some sense?

Author: I just submitted a change with a version of Oliver's suggestion.

> 

Reviewer 3: Github won't let me comment on line 33, but:

> New service worker events received before the idle timer or long running task timer expire would reset the timers and extend the service worker's lifetime.

This should read "... would reset the idle timer and..." vs "reset the timers and".  The long-running task timer is not reset.

Author: Done.

## What's changed

> As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now stay alive as long as they are actively processing events. Be aware that:

Reviewer 3: ```suggestion
As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now typically stay alive as long as they are actively processing events. Be aware that:
```

* The service worker will terminate after 30 seconds if there are no events.

> * The service worker will terminate after 5 minutes if a single request (event or API call) takes longer than 5 minutes to process.

Reviewer 3: ```suggestion
* The service worker will terminate after 5 minutes if a single request, such as an event or API call, takes longer than 5 minutes to process (with the exception of a native messaging connection).
```

You are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally.

> In a future release, we may change how specific events are processed to prevent keeping service workers alive indefinitely, but without shutting down long-running work. In the meantime, make sure your extensions yield when possible and prepare for termination by persisting extension states. 

Reviewer 3: This line doesn't make sense to me.

> to prevent keeping service workers alive indefinitely

We don't mind this as long as it's legitimately useful work.

> but without shutting down long-running work

This seems directly contradictory to the long-running task timer, which explicitly _will_ shut down long-running work.

If you want to keep this sentence, maybe:

"Future releases of Chromium may prevent artificially extending service worker lifetimes."

That said, I we could even just remove this whole sentence, and just append the second to the previous paragraph:

"... you should test your extensions to ensure that they're not doing this unintentionally and make sure your extensions yield whenever possible..."

Author: I took the second approach. The existing language was Simeon's. Not only do I prefer to stay away from remarks about the future, such promises are also contrary to Google's style guide.

As of Chrome 110, all events reset the lidle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now stay alive as long as they are actively processing events.

> Because the service worker will still terminate after five minutes if there are no events, you are still strongly encouraged to design your extensions to be event-based and guard against unexpected termination. Not only should you refrain from keeping your extensions alive indefinitely (which we consider an anti-pattern), you should test your extensions to ensure that they're not doing this unintentionally.

Reviewer 3: This still isn't quite right.
- The service worker will terminate after 30 seconds if there are no events
- The service worker will terminate after 5 minutes if a single request (event or API call) takes longer than 5 minutes to process

Author: PTAL

## What's changed

As of Chrome 110, all events reset the idle timer and the idle timeout will not occur if there are pending events. In other words, assuming there are no unexpected interruptions, extension service workers will now typically stay alive as long as they are actively processing events. In addition, calls to extension specific Chrome APIs, such as `chrome.storage.local.get()`, will reset the idle timeout. Specifically:

* The service worker terminates after 30 seconds of inactivity. (Receiving an event or calling an extension API resets this timer).
* The service worker terminates if a single request, such as an event or API call, takes longer than 5 minutes to process.

Some APIs like native messaging provide a strong keep-alive which cancel both of these timers.

> We are till working to ensure that extension service workers are terminated when possible, without shutting down long-running work. A resource-concious extension service workers should always yield when possible. Additionally, extensions should prepare for unexpected termination by persisting state. This ensures against unpredictable events like the browser being forcefully closed by the user.

Reviewer: ```suggestion
We are still working to ensure that extension service workers are terminated when possible, without shutting down long-running work. Resource-concious extension service workers should always yield when possible. Additionally, extensions should prepare for unexpected termination by persisting state. This guards against unpredictable events like the browser being forcefully closed by the user.
```
